consensus/istanbul/proxy/types.go

```diff
@@ -0,0 +1,183 @@
+//
+// The celo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The celo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the celo library. If not, see <http://www.gnu.org/licenses/>.
+
+package proxy
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus"
+	"github.com/ethereum/go-ethereum/consensus/istanbul"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+var (
+	// errUnauthorizedMessageFromProxiedValidator is returned when the received message expected to be signed
+	// by the proxied validator, but signed from another key
+	errUnauthorizedMessageFromProxiedValidator = errors.New("message not authorized by proxied validator")
+
+	// errInvalidEnodeCertificate is returned if the enode certificate is invalid
+	errInvalidEnodeCertificate = errors.New("invalid enode certificate")
+
+	// errUnauthorizedProxiedValidator is returned if the peer connecting is not the
+	// authorized proxied validator
+	errUnauthorizedProxiedValidator = errors.New("unauthorized proxied validator")
+
+	// ErrStoppedProxyHandler is returned if proxy handler is stopped
+	ErrStoppedProxyHandler = errors.New("stopped proxy handler")
+
+	// ErrStartedProxyHandler is returned if proxy handler is already started
+	ErrStartedProxyHandler = errors.New("started proxy handler")
+
+	// ErrNodeNotProxiedValidator is returned if this node is not a proxied validator
+	ErrNodeNotProxiedValidator = errors.New("node not a proxied validator")
+
+	// ErrNodeNotProxy is returned if this node is not a proxy
+	ErrNodeNotProxy = errors.New("node not a proxy")
+)
+
+type ProxyEngine interface {
+	HandleMsg(peer consensus.Peer, msgCode uint64, payload []byte) (bool, error)
+	RegisterProxiedValidatorPeer(proxiedValidatorPeer consensus.Peer)
+	UnregisterProxiedValidatorPeer(proxiedValidatorPeer consensus.Peer)
+	// SendDelegateSignMsgToProxiedValidator(msg []byte) error
+	SendEnodeCertificateMsgToProxiedValidator(msg *istanbul.Message) error
+	GetProxiedValidatorsInfo() ([]ProxiedValidatorInfo, error)
+}
+
+type ProxiedValidatorEngine interface {
+	Start() error
+	Stop() error
+	AddProxy(node, externalNode *enode.Node) error
+	RemoveProxy(node *enode.Node) error
+	RegisterProxyPeer(proxyPeer consensus.Peer) error
+	UnregisterProxyPeer(proxyPeer consensus.Peer)
+	sendValEnodesShareMsg(proxyPeer consensus.Peer, remoteValidators []common.Address) error
```

Comment:
this is `lowcase`, should this be on the interface?

Reply:
It's needed by the proxy handler, which is in this package.  The proxy handler contains an interface to the proxied validator engine, as opposed to the underlying object.